File: android/src/main/java/com/unistyles/Equatable.kt
package com.unistyles

import com.margelo.nitro.unistyles.Dimensions
import com.margelo.nitro.unistyles.Insets
import com.margelo.nitro.unistyles.UnistyleDependency
import com.margelo.nitro.unistyles.UnistylesNativeMiniRuntime

fun Dimensions.isEqualTo(other: Dimensions): Boolean {
    return this.width == other.width && this.height == other.height
}

fun Insets.isEqualTo(other: Insets): Boolean {
    return this.top == other.top && this.bottom == other.bottom &&
        this.left == other.left && this.right == other.right &&
        this.ime == other.ime
}

fun NativePlatformAndroid.diffMiniRuntimes(lhs: UnistylesNativeMiniRuntime, rhs: UnistylesNativeMiniRuntime): Array<UnistyleDependency> {
    val dependencies: MutableList<UnistyleDependency> = mutableListOf()

    if (lhs.colorScheme != rhs.colorScheme) {
        dependencies.add(UnistyleDependency.COLORSCHEME)
    }

    if (!lhs.screen.isEqualTo(rhs.screen)) {
        dependencies.add(UnistyleDependency.DIMENSIONS)
    }

    if (lhs.screen.width != rhs.screen.width) {
        dependencies.add(UnistyleDependency.BREAKPOINTS)
    }

    // no need to check isLandscape, as it's always opposite
    if (lhs.isPortrait != rhs.isPortrait) {
        dependencies.add(UnistyleDependency.ORIENTATION)
    }

    if (lhs.contentSizeCategory != rhs.contentSizeCategory) {
        dependencies.add(UnistyleDependency.CONTENTSIZECATEGORY)
    }

    if (!lhs.insets.isEqualTo(rhs.insets)) {
        dependencies.add(UnistyleDependency.INSETS)
    }

    if (lhs.fontScale != rhs.fontScale) {
        dependencies.add(UnistyleDependency.FONTSCALE)
    }

    if (!lhs.statusBar.isEqualTo(rhs.statusBar)) {
        dependencies.add(UnistyleDependency.STATUSBAR)
    }

    if (!lhs.navigationBar.isEqualTo(rhs.navigationBar)) {
        dependencies.add(UnistyleDependency.NAVIGATIONBAR)
    }

    // rtl and pixel ratio are not dynamic

    return dependencies.toTypedArray()
}

File: packages/unistyles/ios/Equatable.swift
import Foundation

func dimensionsEqual(_ lhs: Dimensions, _ rhs: Dimensions) -> Bool {
    return lhs.width == rhs.width && lhs.height == rhs.height
}

func insetsEqual(_ lhs: Insets, _ rhs: Insets) -> Bool {
    if (lhs.ime != rhs.ime) {
        return false
    }
    
    return lhs.top == rhs.top && lhs.bottom == rhs.bottom &&
           lhs.left == rhs.left && lhs.right == rhs.right
}

extension UnistylesNativeMiniRuntime {
    static func diff(lhs: UnistylesNativeMiniRuntime, rhs: UnistylesNativeMiniRuntime) -> Array<UnistyleDependency> {
        var dependencies: Array<UnistyleDependency> = []

        if (lhs.colorScheme != rhs.colorScheme) {
            dependencies.append(UnistyleDependency.colorscheme)
        }

        if !dimensionsEqual(lhs.screen, rhs.screen) {
            dependencies.append(UnistyleDependency.dimensions)
        }

        if (lhs.screen.width != rhs.screen.width) {
            dependencies.append(UnistyleDependency.breakpoints)
        }

        // no need to check isLandscape, as it's always opposite
        if (lhs.isPortrait != rhs.isPortrait) {
            dependencies.append(UnistyleDependency.orientation)
        }

        if (lhs.contentSizeCategory != rhs.contentSizeCategory) {
            dependencies.append(UnistyleDependency.contentsizecategory)
        }

        if !insetsEqual(lhs.insets, rhs.insets) {
            dependencies.append(UnistyleDependency.insets)
        }

        if (lhs.fontScale != rhs.fontScale) {
            dependencies.append(UnistyleDependency.fontscale)
        }

        if !dimensionsEqual(lhs.statusBar, rhs.statusBar) {
            dependencies.append(UnistyleDependency.statusbar)
        }

        // rtl and pixel ratio are not dynamic
        // navigation bar is not supported for iOS

        return dependencies
    }
}
